pkg/skaffold/deploy/status_check.go

```diff
@@ -59,49 +59,79 @@ type counter struct {
 	failed  int32
 }
 
-func StatusCheck(ctx context.Context, defaultLabeller *DefaultLabeller, runCtx *runcontext.RunContext, out io.Writer) error {
+// StatusChecker wait for the application to be totally deployed.
+type StatusChecker interface {
+	Check(context.Context, io.Writer) error
+}
+
+// StatusChecker runs status checks for pods and deployments
+type statusChecker struct {
+	// TODO use an interface #4605
+	runCtx          *runcontext.RunContext
+	labeller        *DefaultLabeller
+	deadlineSeconds int
+	muteLogs        bool
+}
+
+// NewStatusChecker returns a status checker which runs checks on deployments and pods.
+func NewStatusChecker(runCtx *runcontext.RunContext, labeller *DefaultLabeller) StatusChecker {
+	return statusChecker{
+		muteLogs:        runCtx.Muted().MuteStatusCheck(),
+		runCtx:          runCtx,
+		labeller:        labeller,
+		deadlineSeconds: runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds,
+	}
+}
+
+// Run runs the status checks on deployments and pods deployed in current skaffold dev iteration.
+func (s statusChecker) Check(ctx context.Context, out io.Writer) error {
 	event.StatusCheckEventStarted()
-	errCode, err := statusCheck(ctx, defaultLabeller, runCtx, out)
+	errCode, err := s.statusCheck(ctx, out)
 	event.StatusCheckEventEnded(errCode, err)
 	return err
 }
 
-func statusCheck(ctx context.Context, defaultLabeller *DefaultLabeller, runCtx *runcontext.RunContext, out io.Writer) (proto.StatusCode, error) {
+func (s statusChecker) statusCheck(ctx context.Context, out io.Writer) (proto.StatusCode, error) {
 	client, err := pkgkubernetes.Client()
 	if err != nil {
 		return proto.StatusCode_STATUSCHECK_KUBECTL_CLIENT_FETCH_ERR, fmt.Errorf("getting Kubernetes client: %w", err)
 	}
 
 	deployments := make([]*resource.Deployment, 0)
-	for _, n := range runCtx.GetNamespaces() {
-		newDeployments, err := getDeployments(client, n, defaultLabeller,
-			getDeadline(runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds))
+	for _, n := range s.runCtx.GetNamespaces() {
+		newDeployments, err := getDeployments(client, n, s.labeller,
+			getDeadline(s.runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds))
 		if err != nil {
 			return proto.StatusCode_STATUSCHECK_DEPLOYMENT_FETCH_ERR, fmt.Errorf("could not fetch deployments: %w", err)
 		}
 		deployments = append(deployments, newDeployments...)
 	}
 
-	deadline := statusCheckMaxDeadline(runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds, deployments)
-
 	var wg sync.WaitGroup
 
 	c := newCounter(len(deployments))
 
+	sCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	for _, d := range deployments {
 		wg.Add(1)
 		go func(r *resource.Deployment) {
 			defer wg.Done()
 			// keep updating the resource status until it fails/succeeds/times out
-			pollDeploymentStatus(ctx, runCtx, r)
+			pollDeploymentStatus(sCtx, s.runCtx, r)
 			rcCopy := c.markProcessed(r.Status().Error())
```

Comment:
Can we check if `sCtx` has been cancelled and return?

Reply:
we did not do that in master. Its a good optimization especially if a user hits ^C early